core/chains/evm/logpoller/log_poller.go

```diff
@@ -366,6 +367,27 @@ func (lp *logPoller) Replay(ctx context.Context, fromBlock int64) error {
 	if fromBlock < 1 || fromBlock > latest.Number {
 		return pkgerrors.Errorf("Invalid replay block number %v, acceptable range [1, %v]", fromBlock, latest.Number)
 	}
+
+	// Backfill all logs up to the current finalized block outside the LogPoller's main loop.
+	latestFinalizedBlockNumber, err := lp.latestFinalizedBlockNumber(ctx)
+	if err != nil {
+		return err
+	}
+	// Do the replay of finalized blocks outside the main loop.
+	if fromBlock <= latestFinalizedBlockNumber {
+		err = lp.backfill(ctx, fromBlock, latestFinalizedBlockNumber)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Backfill everything after latest finalized block in main loop to avoid concurrent writes during reorg
+	// We assume that number of logs between latest finalized and current head is small enough to be processed in main loop
```

Comment:
This assumption is only valid if latestFinalizedBlockNumber were determined by what the chain currently says it is.  But we read it from db... if the node was down for a week or two, the value we have for `latestFinalizedBlockNumber` would also be from a week or two ago. And if the node is starting up for the first time, I notice... `latestFinalizedBlock()` returns 0.

I'm not sure this is actually a problem, because in the main loop it should still do a backfill (I think?), but if you're worried about that overwhelming the main loop it seems like this won't address that case.

Reply:
Oh, nevermind on the last comment I think I get it--the code is right, comment is just confusing because there are two different latestFinalizedBlock numbers being referenced with the same name.

I think we should start using two different names for these, maybe `savedLatestFinalizedBlockNumber` and `latestFinalizedBlockNumber`?  We want to backfill from `fromBlock` up through `savedLatestFinalizedBlockNumber` here, and then backfill from `savedLatestFinalizedBlockNumber` to `lastestFinalizedBlockNumber` in the main loop, then fetch 1-by-1 after that, right?

Reply:
Exactly, I can renamed that to savedFinalizedBlockNumber to reduce confusion